src/sentry/snuba/metrics/extraction.py

```diff
@@ -620,13 +611,37 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
     ]
 
 
-# This is used to map a metric to a function which generates a specification
-_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {
-    "on_demand_failure_count": failure_tag_spec,
-    "on_demand_failure_rate": failure_tag_spec,
-    "on_demand_apdex": apdex_tag_spec,
+DERIVED_METRICS = {
+    "apdex": {
+        "tag_spec_func": apdex_tag_spec,
+        "hash_method": "op_plus_arg",
+    },
+    "failure_count": {
+        "tag_spec_func": failure_tag_spec,
+        "hash_method": "op",
+    },
+    "failure_rate": {
+        "tag_spec_func": failure_tag_spec,
+        "hash_method": "op",
+    },
 }
 
+# This is used to map a metric to a function which generates a specification
+_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {}
+for metric in DERIVED_METRICS.keys():
+    _DERIVED_METRICS[cast(MetricOperationType, f"on_demand_{metric}")] = cast(
```

Comment:
cast removes all type safety and should really only be used for bugs in mypy -- these should go back to being static dictionaries and a test should be added to check they are in sync

same for the other usages of cast